### PR TITLE
RSX: Story about four optimizations

### PR DESCRIPTION
**Pull request by elad335**
* Prevent reload of vertex transform program if program data write did not change the program.
* Optimize sucessive fragment program and vertex program use in two ways:
1. If no change was made to the shader program, try to reuse the compiled program output without heavy validation.
2. If data partially matches between current vertex and shader programs to the previous ones, try to do a fast comparison to see f they match. Skipping hash table resolving.
* Add a method that encapsulates fragment texture changes which is a minor optimization of FIFO execution.

**Metal Gear Online (Menu): ~7000 FPS to 10,000 fps**
**Ninja Guiden Sigma:  194 fps to 202 fps**
**Ratchet And Clank: Toold Of Destruction: 22 -> 24 FPS!**

### CODE REVIEW

**Comment by elad335** [comment]
I have removed some other state-management from storage hashing.

**Comment by cipherxof** [comment]
After https://github.com/RPCS3/rpcs3/pull/16797/commits/55c6f021ea9afc371608e99f4bee6449b51dd8d3 (was about 7,000fps before this commit):

Screenshot From 2025-03-02 22-56-03

image

Seems about the same everywhere else that I've tested compared to https://github.com/RPCS3/rpcs3/pull/16780

**Comment by elad335** [comment]
@cipherxof thanks, pushed more commits which hopefully should eliminate hashing collisions. Please retest on b8dbf3d

**Comment by elad335** [comment]
@cipherxof build is ready!

File: rpcs3/Emu/RSX/Program/ProgramStateCache.cpp [hunk 1]
```diff
@@ -427,13 +427,13 @@ vertex_program_utils::vertex_program_metadata vertex_program_utils::analyse_vert
 
 usz vertex_program_storage_hash::operator()(const RSXVertexProgram &program) const
 {
-	const usz ucode_hash = vertex_program_utils::get_vertex_program_ucode_hash(program);
+	const usz ucode_hash = program.get_sub_hash();
```
Comment (kd-11) [inline rpcs3/Emu/RSX/Program/ProgramStateCache.cpp:430]: Storae hash is used to fingerprint ucode dumps on disk, cannot use subhash here.

File: rpcs3/Emu/RSX/Program/ProgramStateCache.cpp [hunk 2]
```diff
@@ -425,31 +348,41 @@ vertex_program_utils::vertex_program_metadata vertex_program_utils::analyse_vert
...
+{
+	static constexpr u32 s_zero[4]{};
+
+	const usz prog_size = data.size() / 4;
+
+	const void* sdata_base = prog_size ? data.data() : s_zero;
+	const u8* sdata_start = static_cast<const u8*>(sdata_base);
+
+	const u8* sdata_mid1 = sdata_start + (prog_size / 8) * sizeof(v128);
+	const u8* sdata_mid2 = sdata_start + (prog_size / 4) * sizeof(v128);
```
Comment (Whatcookie) [inline rpcs3/Emu/RSX/Program/ProgramStateCache.cpp:361]: You're hashing this data without checking if they're constants or instructions first.

File: rpcs3/Emu/RSX/Program/ProgramStateCache.cpp [hunk 3]
```diff
@@ -425,31 +354,42 @@ vertex_program_utils::vertex_program_metadata vertex_program_utils::analyse_vert
 	return result;
 }
 
+u64 RSXVertexProgram::get_sub_hash() const noexcept
+{
+	static constexpr u32 s_zero[4]{};
+
+	const usz prog_size = data.size() / 4;
+
+	const void* sdata_base = prog_size ? data.data() : s_zero;
```
Comment (kd-11) [inline rpcs3/Emu/RSX/Program/ProgramStateCache.cpp:363]: There are single-instruction shaders btw used to do FB clear or just flood fill

File: rpcs3/Emu/RSX/Program/ProgramStateCache.cpp [hunk 4]
```diff
@@ -718,22 +660,46 @@ usz fragment_program_utils::get_fragment_program_ucode_hash(const RSXFragmentPro
 	return 0;
 }
 
+u64 RSXFragmentProgram::get_sub_hash() const noexcept
+{
+	static constexpr u32 s_zero[4]{};
+
+	const usz prog_size = ucode_length / 16;
+
+	const void* sdata_base = ucode_length ? data.data_ptr : s_zero;
```
Comment (kd-11) [inline rpcs3/Emu/RSX/Program/ProgramStateCache.cpp:669]: Same here. Watch out for very short programs. Min length is 1

**Comment by Whatcookie** [comment]
Latest build is still compiling shaders on every run
image

**Comment by elad335** [comment]
I decided to go on a different route at the end - if the last program cached seems very similar to the current one, do not do hashing against the hash table, instead compare the two programs directly and return them if they match precisely.
The drawback is of course, that fast likely-ness test is not catching all cases of mismatch and thus this may add a second comparison in edge cases. May benefit RaC ToD which does not invalidate FP and VP programs down its pipeline.

I also added a minor optimization for VP UCODE invalidation checks.

**Comment by Whatcookie** [comment]
Provides a nice uplift when avx-512-icl is enabled, and an even bigger uplift for those without.

With AVX-512-ICL (~204->209avg)
image

With AVX-512-ICL paths disabled at compile time (~194->202avg)
image

**Comment by Whatcookie** [comment]
uplift is very good in ratchet tod
image

**Review by kd-11 — DISMISSED** [review]
Nice work.
Do clean up the code as suggested though before merging.

**Comment by cipherxof** [comment]
> I decided to go on a different route at the end - if the last program cached seems very similar to the current one, do not do hashing against the hash table, instead compare the two programs directly and return them if they match precisely. The drawback is of course, that fast likely-ness test is not catching all cases of mismatch and thus this may add a second comparison in edge cases. May benefit RaC ToD which does not invalidate FP and VP programs down its pipeline.
> 
> I also added a minor optimization for VP UCODE invalidation checks.
> 
> @cipherxof can you test it?

Performance is the same as my last test :+1:

**Comment by elad335** [comment]
I have pushed another optimization, this time if VP shader has not been changed since last VP shader fetch we skip the comparison altogether. This is trivial for VP as its changes are placed on the instruction stream.

May benefit RaC ToD and other games.

File: rpcs3/Emu/RSX/Program/ProgramStateCache.h [hunk 5]
```diff
@@ -202,31 +225,52 @@ class program_state_cache
...
 	/// bool here to inform that the program was preexisting.
-	std::tuple<const fragment_program_type&, bool> search_fragment_program(const RSXFragmentProgram& rsx_fp, bool force_load = true)
+	std::tuple<const fragment_program_type&, bool> search_fragment_program(const RSXFragmentProgram& rsx_fp, usz /*rsx_fp_invalidation_count*/)
 	{
 		bool recompile = false;
 		typename binary_to_fragment_program::iterator it;
 		fragment_program_type* new_shader;
-
 		{
+			thread_local const std::pair<const RSXFragmentProgram, fragment_program_type>* prev_fp = nullptr;
```
Comment (kd-11) [inline rpcs3/Emu/RSX/Program/ProgramStateCache.h:234]: Isn't this pointer invalidated if the container has an insert done on it?
This whole approach is q bit too convoluted. Why integer counter instead of bool? Why thread-local instead of atomic? etc.
I get the general idea, but it requires a much more involved implementation to be "proper".

Comment (elad335) [reply rpcs3/Emu/RSX/Program/ProgramStateCache.h:234]: Yep, unordered_map does. I set fetched invalidation count to umax in this case, I can also set to pointer to be null for the intention to be cleaer.

**Comment by elad335** [comment]
Pushed another optimization that reduces FP shader comparison between sequencing draws.

File: rpcs3/Emu/RSX/NV47/HW/common.cpp [hunk 6]
```diff
@@ -72,6 +77,84 @@ namespace rsx
 			}
 		}
 
+		void set_texture_configuration_command(rsx::context* ctx, u32 reg)
```
Comment (kd-11) [inline rpcs3/Emu/RSX/NV47/HW/common.cpp:80]: This is a restructure, not an optimization. I'd suggest moving this to a separate PR to avoid regressions.

File: rpcs3/Emu/RSX/NV47/HW/common.cpp [hunk 7]
```diff
@@ -62,8 +62,13 @@ namespace rsx
 			return vm::cast(get_address(offset, location));
 		}
 
-		void set_fragment_texture_dirty_bit(rsx::context* ctx, u32 index)
+		void set_fragment_texture_dirty_bit(rsx::context* ctx, u32 arg, u32 index, bool /*is_shader_config*/)
 		{
+			if (REGS(ctx)->latch == arg)
```
Comment (kd-11) [inline rpcs3/Emu/RSX/NV47/HW/common.cpp:67]: Caller should check this on their side.

File: rpcs3/Emu/RSX/Program/ProgramStateCache.h [hunk 8]
```diff
@@ -245,10 +246,15 @@ class program_state_cache
 				{
 					if (rsx_fp_invalidation_count != umax && prev_rsx_count == rsx_fp_invalidation_count)
 					{
-						return std::forward_as_tuple(prev_fp->second, true);
+						// Shader UCODE must be the same.
+						// Shader config changes are not tracked at the moment
+						// Compare manually
+						if (program_hash_util::fragment_program_compare::config_only(prev_fp->first, rsx_fp))
```
Comment (kd-11) [inline rpcs3/Emu/RSX/Program/ProgramStateCache.h:252]: Config changes are tracked by the fragment_state_dirty flag. Though I suppose there's no harm in double-checking here since the test is simple.

File: rpcs3/Emu/RSX/NV47/HW/nv47_sync.hpp [hunk 9]
```diff
@@ -27,6 +27,9 @@ namespace rsx
 					return;
 				}
 
+				// Syncronization point, may be associated with memory changes without actually changing addresses
+				RSX(ctx)->m_graphics_state |= rsx::pipeline_state::fragment_program_ucode_dirty;
```
Comment (kd-11) [inline rpcs3/Emu/RSX/NV47/HW/nv47_sync.hpp:31]: Move to caller or obey flush flags. Not all synchronization points mean that CPU is fully synced. e.g Texture read semaphore without SRM should ignore slow activity.
Only FlushPipe does a full sync, FlushDMA does not.

File: rpcs3/Emu/RSX/NV47/HW/common.cpp (hunk 7 above)
Comment (elad335) [reply rpcs3/Emu/RSX/NV47/HW/common.cpp:67]: This was to avoid including rsx_methods.h here.
